### PR TITLE
Implement jest plugin

### PR DESCRIPTION
**Pull request by kibertoad**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -5,7 +5,8 @@
   "main": "lib/index.js",
   "scripts": {
     "test": "mocha",
-    "test:coverage": "nyc npm test",
+    "test:jest": "jest --runInBand --config=jest.config.json",
```
Comment (kibertoad) [inline package.json]: Plugin coverage tests for other plugins explode when executed in jest, hence it needs to be separated.

**Comment by kobik** [comment]
Thanks @kibertoad, this is very nice addition!

Can you make the output of a failed assertion  to look more like this?
image

I'm less familiar with Jest API, but I'm guessing that it can be achieved by using https://jestjs.io/docs/en/expect.html#thisutils

**Comment by kibertoad** [comment]
@kobik Thanks for the feedback! I'll try :D

File: types/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+// TypeScript Version: 3.2
+export {};
+declare global {
```
Comment (kibertoad) [inline types/index.d.ts]: This is needed to augment Jest expect definition so that new methods are recognized by Typescript.

File: package.json [hunk 3]
```diff
@@ -40,32 +45,47 @@
...
   "dependencies": {
     "api-schema-builder": "^1.1.0",
+    "jest-diff": "^24.8.0",
+    "jest-matcher-utils": "^24.8.0",
     "chalk": "^2.4.2",
     "columnify": "^1.5.4",
     "lodash": "^4.17.11"
   },
   "devDependencies": {
+    "@types/jest": "^24.0.15",
```
Comment (kibertoad) [inline package.json]: There is an unfortunate consequence of this being a devDependency - TS compilation will fail for people who use TS but don't use Jest, unless they include Jest typings themselves. Alternative is to move it to prodDependencies, but then people will transiently receive types in runtime as well. I'm not sure how you would want to proceed here.

**Comment by kibertoad** [comment]
@kobik Any updates?

File: types/index.d.ts (hunk 2 above)
Comment (kobik) [reply types/index.d.ts]: this should allow intellisense when adding expectations to tests? if so it doesn't work for me vscode. not sure if this is an issue on my end.

Comment (kibertoad) [reply types/index.d.ts]: Yes, exactly. Works perfectly with WebStorm when consuming as external dependency (as then this file is auto-imported due to reference in package.json).
When consumed from within the project itself, make sure to import this file directly.
Are you trying it in a TS project? Not sure if VSCode supports inferring intellisense for non-TS projects from TS types.

Comment (kibertoad) [reply types/index.d.ts]: Actually, IntelliSense is not even the most important part. When in TypeScript project, tests that are using new assertions wouldn't even compile without this  (due to missing methods on jest `expect`), unless you put `// @ts-ignore` all over the place, which make the feature pretty much unusable in such environment.

Comment (kobik) [reply types/index.d.ts]: Cool!

Weird it's not working for me, though.

I can accept the PR like this, but would it be possible to add this also for chai and should?

Comment (kibertoad) [reply types/index.d.ts]: Appreciate it!
Of course, I can also fully type the configs and validators. Can it happen in a separate PR after this one is merged, though?

Comment (kobik) [reply types/index.d.ts]: Sure, approving

**Pull request merged**